src/client/dfuse/dfuse_thread.c

```diff
@@ -0,0 +1,189 @@
+/**
+ * (C) Copyright 2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#include <pthread.h>
+
+#include <fuse3/fuse_lowlevel.h>
+#define D_LOGFAC DD_FAC(dfuse)
+#include "dfuse.h"
+
+struct dfuse_thread {
+	d_list_t	dt_threads;
+	pthread_t	dt_id;
+	struct fuse_buf dt_fbuf;
+	struct dfuse_tm	*dt_tm;
+};
+
+struct dfuse_tm {
+	d_list_t		tm_threads;
+	pthread_mutex_t		tm_lock;
+	struct fuse_session	*tm_se;
+	sem_t			tm_finish;
+	bool			tm_exit;
+	int			tm_error;
+};
+
+static int
+start_one(struct dfuse_tm *mt);
+
+static void
+*dfuse_do_work(void *arg)
+{
+	struct dfuse_thread	*dt = arg;
+	struct dfuse_tm		*dtm = dt->dt_tm;
+	int rc;
+
+	while (!fuse_session_exited(dtm->tm_se)) {
+		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
+		rc = fuse_session_receive_buf(dtm->tm_se, &dt->dt_fbuf);
+		pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+		if (rc == -EINTR)
+			continue;
+		if (rc <= 0) {
+			if (rc < 0) {
+				fuse_session_exit(dtm->tm_se);
+				dtm->tm_error = -rc;
+			}
+			break;
+		}
+
+		DFUSE_TRA_INFO(dt, "Received %d bytes", rc);
+
+		D_MUTEX_LOCK(&dtm->tm_lock);
```

Comment:
Sanity check...
Could this cause a deadlock? When `dfuse_loop` holds the mutex when it calls `pthread_cancel`, and this thread would not be in a cancelable state. So both would block.
This might happen if some other thread calls `fuse_session_exit`, which would trigger `dfuse_loop` to try to cancel all threads.

But maybe I overlooked something.

Reply:
Does this help:

https://github.com/libfuse/libfuse/blob/master/lib/fuse_lowlevel.c#L3128

Reply:
Hmm. What I'm thinking is:
Thread 1 calls `fuse_session_exit` on line 64.
Thread 2 is somewhere between Line 60 and Line 71.
`dfuse_loop` sees exit=true and gets the lock on Line 165.
`dfuse_loop` calls `pthread_cancel` on Thread 2 on line 167.
Thread 2 is still somewhere between Line 60 and Line 71, so it's not cancelable.
`dfuse_loop` blocks until Thread 2 becomes cancelable.
Thread 2 tries to get the lock on Line 72.
Thread 2 blocks until `dfuse_loop` releases the lock.

Reply:
https://man7.org/linux/man-pages/man3/pthread_cancel.3.html

If a thread has disabled cancellation, then a cancellation request remains queued until the thread enables cancellation

Reply:
Ah, I was reading this:
https://man7.org/linux/man-pages/man3/pthread_setcancelstate.3.html
```
PTHREAD_CANCEL_DISABLE
              The thread is not cancelable.  If a cancellation request is
              received, it is blocked until cancelability is enabled.
```
I was thinking that `pthread_cancel` itself is blocked, but it seems the request is just queued and `pthread_cancel` returns.

Thanks, and sorry!